MongoDB Batch Plugins

gokulavasan opened the pull request:
Build : http://builds.cask.co/browse/CPG-PLUG1

JIRA : https://issues.cask.co/browse/CDAP-3867
- [X] Add unit tests
- [x] Testing on Cluster

gokulavasan committed:
Adding MongoDBBatchSource

gokulavasan committed:
Adding MongoBatchSink

gokulavasan committed:
Naming the sources and sinks right

awholegunch commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/sink/MongoDBBatchSink.java:
```diff
@@ -0,0 +1,104 @@
+ */
+@Plugin(type = "batchsink")
+@Name("MongoDB")
+@Description("CDAP MongoDB Batch Sink converts {@link StructuredRecord} to {@link BSONWritable} and writes to MongoDB")
```
These should not have any links in them; they won't be resolved, and instead you'll just have the raw text...:

```
MongoDB Batch Sink converts a StructuredRecord to a BSONWritable and writes it to a MongoDB.
```

Commit:
Update MongoDBBatchSink.java

Edits.

Commit:
Update MongoDBBatchSource.java

Edits.

Commit:
Update TableSink.java

Minor edit: moving a space from the start of a line to the end of the previous.

awholegunch commented:
I've made edits. Please check my changes...Maybe need to change back to using "CDAP MongoDB Batch Source" in the description so people know it is CDAP's...what do you think?

gokulavasan committed:
Remove id field in MongoBatchSink

nitinmotgi commented:
@gokulavasan UI JSON file is missing with this PR. Would it be possible to add one.

gokulavasan commented:
@nitinmotgi That work is captured in this JIRA https://issues.cask.co/browse/CDAP-3869. I was planning to the UI JSON in a separate PR.

gokulavasan committed:
Adding Mongo realtime sink

gokulavasan committed:
Added CDAP UI json

gokulavasan commented:
@nitinmotgi Added UI json.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+        if (field.getSchema().isNullableSimple()) {
+          return convertValue(input, Schema.Field.of(field.getName(), field.getSchema().getNonNullable()));
+        } else {
+          return null;
```
seems like this should error instead. We could check the schema during configure time

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Name(Properties.SPLITTER_CLASS)
+    @Nullable
+    @Description("The name of the Splitter class to use.")
```
what happens if this is not given?

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Name(Properties.INPUT_FIELDS)
+    @Nullable
+    @Description("A projection document limiting the fields that appear in each document.")
```
not sure what a 'projection document' is, is this mongo terminology? Let's add to the description what happens if this is not given.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/realtime/sink/MongoDBRealtimeSink.java:
```diff
@@ -0,0 +1,97 @@
+    for (StructuredRecord record : iterable) {
+      Document document = new Document();
+      for (Schema.Field field : record.getSchema().getFields()) {
+       document.append(field.getName(), record.get(field.getName()));
```
nit: add another space

albertshau commented on mongodb-plugins/src/test/java/co/cask/plugin/etl/testclasses/DataGeneratorSource.java:
```diff
@@ -0,0 +1,151 @@
+@Plugin(type = "realtimesource")
+@Name("DataGenerator")
+@Description("Source that can generate test data for Real-time Stream and Table Sinks.")
+public class DataGeneratorSource extends RealtimeSource<StructuredRecord> {
```
is this added twice? seems to be 2 DataGeneratorSource

albertshau commented:
some small comments, otherwise lgtm

gokulavasan committed:
Addressing PR comments

gokulavasan commented:
@albertshau Thanks for the review. Addressed comments. Please take a look.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+      case ENUM:
+        return input;
+      case ARRAY:
+        return GSON.fromJson(input, List.class);
```
this will ignore the component type of the array.  For example, array of ints and array of strings we get decoded in the same way.

albertshau replied:
does mongo support array fields? If so, does it support arrays of arbitrary type?

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+      case ARRAY:
+        return GSON.fromJson(input, List.class);
+      case MAP:
+        return GSON.fromJson(input, Map.class);
```
similar comment about maps, the key and value types are lost here.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+      case MAP:
+        return GSON.fromJson(input, Map.class);
+      case UNION:
+        if (field.getSchema().isNullableSimple()) {
```
this means nullable arrays and maps are not allowed. Also, if input is null, this will end up in a NPE.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+        return GSON.fromJson(input, Map.class);
+      case UNION:
+        if (field.getSchema().isNullableSimple()) {
+          return convertValue(input, Schema.Field.of(field.getName(), field.getSchema().getNonNullable()));
```
could change convertValue to take a Schema rather than a Field. Then you don't need to create a new object here.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+          return convertValue(input, Schema.Field.of(field.getName(), field.getSchema().getNonNullable()));
+        }
+    }
+    throw new IOException(String.format("Unsupported schema : %s for field '%s'", field.getSchema(), field.getName()));
```
we should do schema validation at configure time to make sure this never happens at runtime. For example, check union fields, check that there are no record fields.

albertshau commented on mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java:
```diff
@@ -0,0 +1,187 @@
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.valueOf(input);
```
how come we need to convert everything to a string and then convert back? does BSONObject not return the correct type?

albertshau commented:
missed some things on the first pass, questions about the record conversion logic.

albertshau commented:
there is a RecordConverter class in cdap-etl-lib which might be useful

gokulavasan committed:
Creating BSON Converter

gokulavasan committed:
Addressing PR comment. Using RecordConverter from cdap-etl-common

gokulavasan commented:
@albertshau Please review. I am using the RecordConverter class from cdap-etl-common. I can move the class to regular package instead of under co.cask.cdap.etl.common once https://github.com/caskdata/cdap/pull/4403 is merged.

gokulavasan committed:
Adding Javadoc

albertshau commented on mongodb-plugins/src/main/java/co/cask/cdap/etl/common/BSONConverter.java:
```diff
@@ -0,0 +1,45 @@
+  private final Schema schema;
+  private final org.apache.avro.Schema avroSchema;
+
+  public BSONConverter(String schemaString) throws IOException {
```
think its a bit cleaner to pass in the schema object rather than parse it here.  Will encourage better error handling and the constructor won't need an exception.

gokulavasan replied:
I had to derive both our schema and avro schema. Hence I had passed in the schemaString.

albertshau commented:
minor comment, plus it would be good to add test cases that use more complex data types.  arrays and maps are being handled differently now than earlier in the pr.

gokulavasan committed:
Merging develop

gokulavasan committed:
Changes to test

gokulavasan committed:
Initial BSON conversion

gokulavasan committed:
Merge remote-tracking branch 'origin/develop' into feature/mongodb-plugin

gokulavasan committed:
Address PR comments

gokulavasan committed:
Validate input schema

gokulavasan committed:
Depend on 3.2.1

gokulavasan commented:
@albertshau Addressed comments. Please take a look again! Thanks!

albertshau commented on mongodb-plugins/src/test/java/co/cask/plugin/etl/test/MongoDBTest.java:
```diff
@@ -0,0 +1,282 @@
+    scanner.close();
+    Assert.assertArrayEquals("AAPL".getBytes(), row1.getRow());
+    Assert.assertEquals(10, (int) row1.getInt("num"));
+    Assert.assertEquals(23.23, row1.getDouble("price"), 0.00001);
```
should also add a check for the "agents" field

gokulavasan replied:
RecordPutTransformer handles only simple types and "agents" field is not a simple type. So the sink schema does not contain that type.

albertshau replied:
ok i see. is the array type tested at all?

gokulavasan replied:
Yes it is, otherwise conversion of BSONObject of StructuredRecord would have failed. I am adding another test from Mongo to Mongo copy that will provide more confidence.

albertshau replied:
ah thats right, got it

albertshau commented:
1 comment, then lgtm

albertshau commented on mongodb-plugins/src/main/java/co/cask/cdap/etl/common/BSONConverter.java:
```diff
@@ -0,0 +1,89 @@
+
+  private Object extractValue(Object object, Schema schema) {
+    Schema.Type fieldType = schema.getType();
+    switch (fieldType) {
```
what happened to handling of nullables?

albertshau commented:
one question about nullable types in the source, otherwise lgtm

gokulavasan committed:
Address PR comment : Add ability to handle nullable schemas

gokulavasan commented:
@albertshau Thanks for the catch! Fixed it (also adjusted the unit test to have a nullable field). Please review.

albertshau commented:
one more comment, plus travis failed

gokulavasan committed:
Addressing PR comment

gokulavasan committed:
Revert changes to when MongoDB is started

gokulavasan committed:
Dummy commit to trigger build

gokulavasan committed:
Javadoc and newline

gokulavasan committed:
Merge remote-tracking branch 'origin/develop' into feature/mongodb-plugin

Commit:
Merge branch 'develop' into feature/mongodb-plugin

gokulavasan committed:
Merge remote-tracking branch 'origin/develop' into feature/mongodb-plugin

gokulavasan committed:
Merge branch 'feature/mongodb-plugin' of http://github.com/caskdata/hydrator-plugins into feature/mongodb-plugin

gokulavasan committed:
Fixing checkstyle, increasing perm space

gokulavasan commented:
@albertshau Thanks for the review. Addressed comment. Travis build is passing now.

nitinmotgi commented:
Please add README.rst format same as https://github.com/caskdata/hydrator-plugins/blob/develop/transform-plugins/README.rst before merging.

albertshau commented:
lgtm after adding a readme like nitin suggested

gokulavasan committed:
Adding README.rst and plugin JSON files

gokulavasan commented:
@albertshau Included README.rst. Please review. Thanks!

albertshau commented:
lgtm
